synapse/util/batching_queue.py

```diff
@@ -0,0 +1,137 @@
+
+        # The currently pending batch of values by key, with a Deferred to call
+        # with the result of the corresponding `process_items` call.
+        self._next_values = {}  # type: Dict[Hashable, List[Tuple[V, defer.Deferred]]]
+
+        # The function to call with batches of values.
+        self.process_items = process_items
+
+        LaterGauge(
+            "synapse_util_batching_queue_number_queued",
+            "The number of items waiting in the queue across all keys",
+            labelnames=("name",),
+            caller=lambda: sum(len(v) for v in self._next_values.values()),
+        )
+
+        LaterGauge(
+            "synapse_util_batching_queue_number_of_keys",
+            "The number of distinct keys that have items queued",
+            labelnames=("name",),
+            caller=lambda: len(self._next_values),
+        )
+
+    async def add_to_queue(self, value: V, key: Hashable = ()) -> R:
+        """Adds the value to the queue with the given key, returning the result
+        of the processing function for the batch that included the given value.
+        """
+
+        # First we create a defer and add it and the value to the list of
+        # pending items.
+        d = defer.Deferred()
+        self._next_values.setdefault(key, []).append((value, d))
+
+        # If we're not currently processing the key fire off a background
+        # process to start processing.
+        if key not in self._processing_keys:
+            run_as_background_process(self._name, self._process_queue, key)
+
+        return await make_deferred_yieldable(d)
+
+    async def _process_queue(self, key: Hashable) -> None:
+        """A background task to repeatedly pull things off the queue for the
+        given key and call the `self.process_items` with the values.
+        """
+
+        try:
+            if key in self._processing_keys:
+                return
+
+            self._processing_keys.add(key)
+
+            while self._next_values:
+                # We purposefully wait a reactor tick to allow us to batch
+                # together requests that we're about to receive. A common
+                # pattern is to call `add_to_queue` multiple times at once, and
+                # deferring to the next reactor tick allows us to batch all of
+                # those up.
+                await self._clock.sleep(0)
+
+                next_values = self._next_values.pop(key, [])
+
+                try:
+                    values = [value for value, _ in next_values]
+                    results = await self.process_items(values)
+
+                    for _, deferred in next_values:
+                        with PreserveLoggingContext():
+                            deferred.callback(results)
+
+                except Exception as e:
+                    for _, deferred in next_values:
+                        with PreserveLoggingContext():
+                            deferred.errback(e)
```

Comment:
Could this cause a Deferred to have both `callback` and `errback` called on it if e.g. the second `deferred.callback` causes an error to be raised?

(I wonder if we should be guarding with `if not deferred.called`?)

Reply:
I don't think `deferred.callback` should ever explode, but may as well add the guard.

Reply:
I would hope not. Another option would be to move the callback calls to an `else` so that only calling the processing function is in the try clause?

Reply:
I think I prefer it as it is, as then we really should always call either `.errback` or `.callback` for each?

Reply:
Right, isn't that what my suggestion guarantees though?

```python
try:
    result = processing_callback(values)
except Exception as e:
    for d in deferreds:
        d.errback(e)
else:
    for d in deferreds:
        d.callback(result)
```

Reply:
But if `d.callback` never throws then that's the same as

```python
try:
    result = processing_callback(values)

    for d in deferreds:
        d.callback(result)
except Exception as e:
    for d in deferreds:
        d.errback(e)
```

?

Reply:
Yes. 😄 

If it *does* throw it has the benefit of causing a stack trace and letting us know about it.

Reply:
OK, you can ignore me. It can't throw (for the reason I thought it could):

https://github.com/twisted/twisted/blob/cae365213868e8226a055f6b5e4d2c0c89c9ab3b/src/twisted/internet/defer.py#L851-L871